File: handy-python/quiz/practice.py
import unittest


def cross_product(origin, point_a, point_b):
    # Your function code here
    x = (point_a[0] - origin[0]) * (point_b[1] - origin[1]) - \
        (point_a[1] - origin[1]) * (point_b[0] - origin[0])
    return x


def area(points: list):
    points.sort()
    if len(points) <= 1:
        return points
    left_turn = []
    for p in points:
        while len(left_turn) >= 2 and cross_product(left_turn[-2], left_turn[-1], p) <= 0:
            left_turn.pop()
        left_turn.append(p)

class TestCrossProduct(unittest.TestCase):

    def test_left_turn(self):
        # Positive result: Counter-clockwise
        o, a, b = (0, 0), (1, 0), (0, 1)
        self.assertGreater(cross_product(o, a, b), 0)

    def test_left_turn(self):
        # Negative result: Clockwise
        o, a, b = (0, 0), (0, 1), (1, 0)
        self.assertLess(cross_product(o, a, b), 0)

    def test_collinear(self):
        # Zero result: Straight line
        o, a, b = (0, 0), (1, 1), (2, 2)
        self.assertEqual(cross_product(o, a, b), 0)

    def test_origin_overlap(self):
        # Edge case: Points are the same
        o = (5, 5)
        self.assertEqual(cross_product(o, o, o), 0)


def calculate_area(points):
    n = len(points)
    area = 0.0
    for i in range(n):
        j = (i + 1) % n
        area += points[i][0] * points[j][1]
        area -= points[j][0] * points[i][1]
    return abs(area) / 2.0


points = [(0, 0), (4, 0), (4, 3), (0, 3)]
print(f"Area: {calculate_area(points)}")

if __name__ == '__main__':
    unittest.main()

File: handy-js/practice.js
function crossProduct(origin, a, b) {
    return (a[0] - origin[0]) * (b[1] - origin[1]) - (a[1] - origin[1]) * (b[0] - origin[0]);
}
function convexHull(points) {
    const sorted = [...points].sort((p1, p2) => (p1[0] === p2[0] ? p1[1] - p2[1] : p1[0] - p2[0]));
    if (sorted.length <= 1)
        return sorted;
    const lower = [];
    for (const p of sorted) {
        while (lower.length >= 2 && crossProduct(lower[lower.length - 2], lower[lower.length - 1], p) <= 0) {
            lower.pop();
        }
        lower.push(p);
    }
    const upper = [];
    for (let i = sorted.length - 1; i >= 0; i--) {
        const p = sorted[i];
        while (upper.length >= 2 && crossProduct(upper[upper.length - 2], upper[upper.length - 1], p) <= 0) {
            upper.pop();
        }
        upper.push(p);
    }
    lower.pop();
    upper.pop();
    return [...lower, ...upper];
}
function calculateArea(points) {
    let area = 0;
    for (let i = 0; i < points.length; i++) {
        const j = (i + 1) % points.length;
        area += points[i][0] * points[j][1];
        area -= points[j][0] * points[i][1];
    }
    return Math.abs(area) / 2;
}
const polygon = [
    [0, 0],
    [4, 0],
    [4, 3],
    [0, 3],
];
console.log(`Area: ${calculateArea(polygon)}`);
